# Functions

Function - a part of program code that can be called from other place in the program

In Python functions declare by key word 'def'. 

In [1]:
def say_hello(): #It's simple print hello to the screen
    print ("Hello")
    
say_hello()

Hello


In [2]:
def mod(a, b): #It's calculate a modulo b. To return result from function we need write key word 'return'
    return a % b

print (mod(3,4))

3


In [3]:
def wrong_mod(a, b): #previous example without return.
    a % b

print (wrong_mod(3,4))

None


The function can return multiple values. In this case it return them as tuple. 

In [4]:
def s_pats(string, pats): #This function search for substrings in pats and when it finds one it return it
    ind = -1
    for p in pats:
        ind = string.find(p)
        if ind != -1:
            break
    return ind, p

res = s_pats("Hello, world", ["Hell", "world",
                              "happiness"])
print (type(res))
ind, p = res
print(ind, p)

&lt;class &#39;tuple&#39;&gt;
0 Hell


And there is examples of fuctions that implements calculation of factorial and N-th Fibonacci number

In [5]:
def factorial(n):
    acc = 1
    for i in range(2,n + 1):
        acc *= i
    return acc

def fibonacci(n):
    f0, f1 = 0, 1
    for i in range(n):
        f0, f1 = f1, f0 + f1
        # temp = f0
        # f0 = f1
        # f1 = f1 + temp
    return f0

In [6]:
print (factorial(5))
print (fibonacci(5))

120
5


## Recursion

Try to ask google)

![recursion](https://res.cloudinary.com/dhbo6ilxb/image/upload/v1541158960/recursion.png)

In Python there implements recursion mechanisms - when the function call to itself. Let's made examples of function above with recursion mechanism.

In [12]:
def factorial_rec(n):
    if n <= 1:
        return 1
    return n * factorial_rec(n - 1) #call to itself

def fibonacci_rec(n):
    if n < 0:
        return 0
    
    if n == 0:
        return 0
    if n == 1:
        return 1
    
    return fibonacci_rec(n - 1) + fibonacci_rec(n - 2)

In [8]:
print (factorial_rec(5))
print (fibonacci_rec(5))

120
5


So, how does it work? It will be usefull for work with this mechanism. When the function meet the call of itself it suspend of calculation of itself, run new version of itself and waits until the new version of it finishes calculations with other parameters. And the new version calls another one version utill one of them calls a version in which are no function calls (in example above the lines with n == 0 or 1). After that remaining functions continue to works one after another. Let's try modified one of examples above.

In [10]:
def factorial_rec_print(n):
    print ("called fact with n = " + str(n))
    if n <= 1:
        return 1
    print ("pause fact with n = " + str(n))
    res = n * factorial_rec(n - 1)
    print ("resume fact with n = " + str(n))
    return res

print (factorial_rec_print (6))

called fact with n = 6
pause fact with n = 6
called fact with n = 5
pause fact with n = 5
called fact with n = 4
pause fact with n = 4
called fact with n = 3
pause fact with n = 3
called fact with n = 2
pause fact with n = 2
called fact with n = 1
resume fact with n = 2
resume fact with n = 3
resume fact with n = 4
resume fact with n = 5
resume fact with n = 6
720


![factorial](https://res.cloudinary.com/dhbo6ilxb/image/upload/v1541159011/recursion_stack_2.png)

### Slowdown when using recursion

When you use recurtion you need additional resources to suspend, calling and resuming of fuctions. 

In [11]:
%%timeit
factorial(1000)

348 µs ± 607 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [13]:
%%timeit
factorial_rec(1000)

517 µs ± 7.97 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


348 vs 517 µs for my computer... And there is worse

In [14]:
%%timeit
fibonacci(20)

1.59 µs ± 5.73 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [15]:
%%timeit
fibonacci_rec(20)

4.12 ms ± 55.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


There are two big problems. Firstly we need to suspend functions when calculated others. Second one is that made one calculation many times (the number of calling of fib(5) in fib (10) for example). We can figth with them... but not today.



![fibonacci](https://res.cloudinary.com/dhbo6ilxb/image/upload/v1541164676/fibonacci_recursion.png)

### Infinite recursion

If you don't made recursion exit conditions we get an error that recursion is too big (inifite actualy)

In [16]:
def factorial_rec_no_exit(n):
    return n * factorial_rec_no_exit(n - 1)

factorial_rec_no_exit(10) # good night, sweet prince

RecursionError: maximum recursion depth exceeded

## Default arguments

Quite often you can meet the situation when you know what arguments can be taken if the user has not specified them (for example sort function where you can choose the algorithm and if user not want to choose it you just write that default is qsort). In that situation you can write the default arguments in the describtion for you function. But there is a special mechanism in Python called default arguments. 

In [17]:
import random

def make_random_sequence(length=200,
                         alphabet="ATGC"):
    seq_lst = [] 
    for i in range(length):
        seq_lst.append(random.choice(alphabet))
    seq = "".join(seq_lst) 
    return seq

In [18]:
make_random_sequence()

&#39;ATCATGTCGGACATACAGGAGCGTCAACACAAGCTGAGCTGTGATCTAGCGTAGCTCTCCGACCTAGCACCGCTATATTAGGTGACCCACTCTTATCGTACTATAGTACTCGCTGCTTCTACGCGATTCTCCGTTAGGTTGTTTCGGTGAACAACCCGACATAATCGAGATGGCGCCGTACGAAGCAGACGTAATCACTG&#39;

In [21]:
make_random_sequence(length=100)

&#39;TGGATGAAATAGCATTTTACAATTGTGAAGTGACAAACAATCCATGACTTTGGATACCTTTCTCATCAACCGGAGCGTTTCGCCCATTGTTGAGAGCAGG&#39;

In [22]:
make_random_sequence(alphabet="AUGC", length=100)

&#39;ACCUGGUAAUCUAACGUUAGUACGGUCGUCGCCUACGCAAGAAGCCCUGAUCCCUUUAACACUACUUUCGGGAAAUUCUCUUACACUCAAACAGGGCAGA&#39;

Or may be you just want to show user how your function work. 

In [23]:
def login(username="anonymous", password=None):
    """Some action"""
    pass

# we can call function in different ways
login("root", "ujdyzysqgfhjkm") 
login("guest")
login()
# Also you can specify the name of argument
login(password="nobody@mail.com") 

You can combine default arguments with usual one. But usual arguments must be declared firstly.

In [24]:
def write_random_fasta(out_file_path, 
                       name="random", 
                       length=200,
                       alphabet="ATGC"):
    out_file = open(out_file_path, "w") # it is better to use with-construction here
    seq = make_random_sequence(length=length, alphabet=alphabet)
    out_file.write(">{}\n".format(name))
    out_file.write("{}\n".format(seq))
    out_file.close()

In [25]:
write_random_fasta("random.fasta")

In [26]:
write_random_fasta()

TypeError: write_random_fasta() missing 1 required positional argument: &#39;out_file_path&#39;

In [27]:
write_random_fasta("random2.fasta", length=10)

In [28]:
def add_to_list(el, lst = []):
    lst.append(el)
    return lst

And now some fun)

In [29]:
print (add_to_list(5, [1,2,3])) # OK
print (add_to_list(5, [])) # OK
print (add_to_list(5)) # OK
print (add_to_list(5)) # WHAT???

[1, 2, 3, 5]
[5]
[5]
[5, 5]


![surprise](https://i.makeagif.com/media/10-25-2018/TA2lrY.gif)

It happens because when Python create a default object it not delete after the funcion end (it keeps in memory untill garbage collector come to us). And when we call the function second time it use the same list. To prevent this just not use mutable objects in default arguments (like a containers or structures where you want change some fields) instead use None.

In [30]:
def add_to_list_wsmf(el, lst = None):
    if lst == None:
        lst = []
    lst.append(el)
    return lst

In [31]:
print (add_to_list_wsmf(5, [1,2,3])) # OK
print (add_to_list_wsmf(5, [])) # OK
print (add_to_list_wsmf(5)) # OK
print (add_to_list_wsmf(5)) # Still OK

[1, 2, 3, 5]
[5]
[5]
[5]
